trecs/models/mf.py

```diff
@@ -0,0 +1,293 @@
+            num_latent_factors,
+        )
+        if model_params is None:
+            model_params = {}
+        num_features_vals = non_none_values(
+            model_params.get("features"), num_latent_factors, num_attributes
+        )
+        if len(num_features_vals) > 1:
+            raise ValueError("Number of latent factors is not the same across inputs.")
+
+        self.num_latent_factors = num_attributes
+        self.model_params = model_params
+        self.als_model = None  # initialize empty model
+        self.all_interactions = pd.DataFrame(columns=["user", "item"])  # empty interactions matrix
+
+        # generate user and item representations as needed
+        if user_representation is None:
+            user_representation = np.zeros((num_users, num_attributes))
+        if item_representation is None:
+            item_representation = Generator(seed=seed).binomial(
+                n=1, p=0.5, size=(num_attributes, num_items)
+            )
+        # if the actual item representation is not specified, we assume
+        # that the recommender system's beliefs about the item attributes
+        # are the same as the "true" item attributes
+        if actual_item_representation is None:
+            actual_item_representation = np.copy(item_representation)
+
+        measurements = [MSEMeasurement()]
+
+        super().__init__(
+            user_representation,
+            item_representation,
+            actual_user_representation,
+            actual_item_representation,
+            num_users,
+            num_items,
+            num_items_per_iter,
+            probabilistic_recommendations=probabilistic_recommendations,
+            measurements=measurements,
+            verbose=verbose,
+            seed=seed,
+            **kwargs
+        )
+
+    def _update_internal_state(self, interactions):
+        """
+        At each training timestep, we keep track of the user/item interactions.
+        """
+        user_item_ids = tuple(zip(self.users.user_vector, interactions))
+        interaction_df = pd.DataFrame(user_item_ids, columns=["user", "item"])
+        self.all_interactions = self.all_interactions.append(interaction_df, ignore_index=True)
+
+    def update_predicted_scores(self):
+        """
+        Calculates predicted scores for every user-item pair by first running an
+        implicit MF algorithm on the interaction data. A glorified wrapper function,
+        but necessary to keep the functionality in BaseRecommender working.
+        """
+        # if no interactions have happened yet, do a simple dot product
+        if self.all_interactions.size == 0:
+            super().update_predicted_scores()
+        else:
+            self.fit_mf()
+
+    def run(
+        self,
+        timesteps=50,
+        startup=False,
+        train_between_steps=False,
+        random_items_per_iter=0,
+        vary_random_items_per_iter=False,
+        repeated_items=True,
+        no_new_items=False,
```

Comment:
Why wouldn't no new items be `True` here?

Reply:
In my mind, the sensible defaults for MF were:
- No new items during the `startup_and_train` phase (so the function signature looks like `def startup_and_train(self, timesteps=50, no_new_items=True)`. I chose to do this because in my mind MF is usually trained on a static dataset where new items aren't constantly being added.
- New items *are* allowed during the `run` phase if content creators are part of the simulation. If I'm a user of the library and I initialize `mf = ImplicitMF(creators=c); mf.startup_and_train(30); mf.run(100)`, I probably expect the content creators to be creating content by default

Reply:
Got it. I didn't catch that the only way new items are added is through content creators, but that makes more sense now.